helloworld-ws quickstart

lnewson opened the pull request:
Hey guys,

This is just a more basic version using JAX-WS that Misty had asked me to spin up as she felt that was missing.

Cheers Lee

Commit:
Added the basic hello world JAX-WS quickstart

sgilda commented:
Hi Lee, thanks for contributing a quickstart! I have a couple of minor comments.
1. I followed the instructions in the README file and ran into a problem right away starting the server. The directions say to "navigate to the root of the JBoss server directory" and type the following command: 'JBOSS_HOME/bin/standalone.sh --server-config=standalone-preview.xml'. When I do that, I get this error:
   java.lang.IllegalStateException: JBAS014805: Could not get main file: standalone-preview.xml. Specified files must be relative to the configuration dir: /home/sgilda/tools/jboss-as-7.1.1.Final/standalone/configuration
   I searched the jboss-as-7.1.1 and jboss-eap-6.0-ER7 directories for that file and don't see anything with that name. 
   What release were you testing?
   Anyway, I just started the server normally with bin/standalone.sh and was able to successfully deploy the quickstart..
2. Since all of the other quickstarts that deploy with the 'mvn clean package jboss-as:deploy' have an 'Access the Application' section noting  "The application will be running at the following URL: http://localhost:8080/jboss-as-helloworld-ws, do you want to mention what this deploy does? Is it needed to run the Arquilliain tests? Even though you don't say to access the application, I'm guessing people will try, so you might want to explain that you see the SOAP XML "No such operation: / (HTTP GET PATH_INFO: /jboss-as-helloworld-ws/)" message. 

Pete, do you want to do a code review, or do you want designate that to someone else?

Thanks, Sande

sgilda commented:
Lee responded on IRC that he wiil do the updates to the README in the next few days.

Commit:
Fixed the README and changed the title to match in web.xml

lnewson commented:
Fixed those issues up Sande. Sorry it took longer then I said but I got caught up with the new version of the CCMS. If you find anything else let me know.

Cheers Lee

sgilda commented:
No problem. It turns out it's too late to make EAP 6 GA anyway. :-(

On 05/21/2012 05:56 PM, lnewson wrote:

> Fixed those issues up Sande. Sorry it took longer then I said but I got caught up with the new version of the CCMS. If you find anything else let me know.
> 
> Cheers Lee
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jbossas/quickstart/pull/268#issuecomment-5834850

sgilda commented:
Hi Lee,
I tested your new instructions and they are much clearer. Thanks! 
Pete, I know you're busy at the moment, but after things die down, would you mind doing a code review?
Thanks,
Sande

pmuir commented on helloworld-ws/pom.xml:
```diff
@@ -0,0 +1,199 @@
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.10</version>
```
If you use the -with-tools BOM from http://www.jboss.org/jdf/stack/jboss-bom/jboss-javaee-6.0-with-tools/ you can then omit the version here.

pmuir commented on helloworld-ws/src/main/java/org/jboss/as/quickstarts/wshelloworld/HelloWorldService.java:
```diff
@@ -0,0 +1,59 @@
+ * @author lnewson@redhat.com
+ */
+
+@WebService(targetNamespace="http://www.jboss.com/jbossas/quickstarts/wshelloworld/HelloWorld")
```
This is certainly the wrong namespace to use, jboss.com has been retired. Do we have guidelines about what namespaces should be used for web services? I guess ideally we would use a URN here, or possibly an example domain like acme.

pmuir commented on helloworld-ws/pom.xml:
```diff
@@ -0,0 +1,199 @@
+                            <outputDirectory>deployments</outputDirectory>
+                            <warName>ROOT</warName>
+                            <!-- Java EE 6 doesn't require web.xml, Maven needs to catch up! -->
+                            <failOnMissingWebXml>false</failOnMissingWebXml>
```
This is wrong, you provide a web.xml

pmuir commented on helloworld-ws/pom.xml:
```diff
@@ -0,0 +1,199 @@
+                <version>2.1.1</version>
+                <configuration>
+                    <!-- Java EE 6 doesn't require web.xml, Maven needs to catch up! -->
+                    <failOnMissingWebXml>false</failOnMissingWebXml>
```
This is wrong, as you do provide a web.xml.

pmuir commented on helloworld-ws/src/test/java/org/jboss/as/quickstarts/wshelloworld/ClientTest.java:
```diff
@@ -0,0 +1,69 @@
+    private static HelloWorldService client;
+    
+    @BeforeClass
+	public static void setup() {
```
Whitespace gone wrong here.

pmuir commented on helloworld-ws/src/test/java/org/jboss/as/quickstarts/wshelloworld/ClientTest.java:
```diff
@@ -0,0 +1,69 @@
+ * 
+ * @author lnewson@redhat.com
+ */
+public class ClientTest {
```
This would work very well as an Arquillian test, as it would sort out deploying the app to JBoss AS for you.

pmuir commented:
In general, this looks good. You need to run the source formatter over it, and the test should be an arquillian test, to deploy the app to the server before the test is run.

lnewson commented:
Thanks for the comments Pete. Also thanks for the info about Open Shift, tbh I just copied that as I didn't know how it worked at at the time of submitting this was pressed for time. I'll go over the comments you made and make the changes.

Cheers Lee

pmuir commented:
Sorry the review as slow!

Have you tested the OpenShift bit?

sgilda commented:
Hi Lee, 

I know you've been busy, but I'm just curious where this stands.

Thanks,
Sande

Commit:
Basic Updates to fix formatting and other issues pointed out by Pete
Muir.

Commit:
Added Arquillian Test Cases to demonstrate a client application example
without having to have an Application Server already running.

Commit:
Updated the README to go with the changes made and to work with the
latest version of OpenShift.

Commit:
Fixed an issue with deploying to the root context in the openshift maven
profile. Removed all old tests and tweaked the arquillian tests.

lnewson commented:
Hey Sande and Pete,

Sorry it took me so long to get around to fixing this, I just was having trouble finding the time and motivation to learn how to use Arquillian. Anyways, I believe I've fixed everything Pete mentioned and the tests are all done using Arquillian now. I've also updated the instructions somewhat since OpenShift has made a few changes since I originally did the quickstart. If you need anything else let me know and I'll make sure I get it done a lot faster this time.

Cheers Lee

sgilda commented:
Hi Lee,

I took a brief look at your quickstart and have a few minor comments:

For the Arquillian tests, you only need to tell them how to run the managed tests. You can copy the section from the wsat-simple/README.md file if you like. We removed remote testing instructions from the individual quickstart README files to make them less complicated. 

It's also nice to describe what they should see in the console for the Arquillian test results. Again, if you look at the current wsat-simple example, this is detailed under a section called "Investigate the Console Output". You can show only the INFO lines if you want: 

INFO: Dynamically creating request wrapper Class org.jboss.as.quickstarts.wshelloworld.jaxws.SayHello
Oct 11, 2012 10:11:57 AM com.sun.xml.internal.ws.model.RuntimeModeler getResponseWrapperClass
INFO: Dynamically creating response wrapper bean Class org.jboss.as.quickstarts.wshelloworld.jaxws.SayHelloResponse
Oct 11, 2012 10:11:57 AM com.sun.xml.internal.ws.model.RuntimeModeler getRequestWrapperClass
INFO: Dynamically creating request wrapper Class org.jboss.as.quickstarts.wshelloworld.jaxws.SayHelloToName
Oct 11, 2012 10:11:57 AM com.sun.xml.internal.ws.model.RuntimeModeler getResponseWrapperClass
INFO: Dynamically creating response wrapper bean Class org.jboss.as.quickstarts.wshelloworld.jaxws.SayHelloToNameResponse
Oct 11, 2012 10:11:57 AM com.sun.xml.internal.ws.model.RuntimeModeler getRequestWrapperClass
INFO: Dynamically creating request wrapper Class org.jboss.as.quickstarts.wshelloworld.jaxws.SayHelloToNames
Oct 11, 2012 10:11:57 AM com.sun.xml.internal.ws.model.RuntimeModeler getResponseWrapperClass
INFO: Dynamically creating response wrapper bean Class org.jboss.as.quickstarts.wshelloworld.jaxws.SayHelloToNamesResponse

In the OpenShift instructions, I see a few places that still have "helloworldrs" instead of "helloworldws". Those are easy to miss. :-)

I haven't looked at the code or pom file, or tested with OpenShift yet but will try to get to those later today.

Thanks,
Sande

sgilda commented:
Hi Lee,

I took a look at your pom.xml file. Things have changed since your created this quickstart.  Sorry about that. BOMs were created to eliminate the need for version numbers hard-coded throughout the file. Now you define properties for

Take a look at one of the latest pom files here for the latest changes: https://github.com/jboss-jdf/jboss-as-quickstart/, for example, the wsat-simple/pom.xml file.

Basically, these are the changes you need to make:
1. The version has changed from 7.1.1-SNAPSHOT tp  7.1.2-SNAPSHOT
2. Add the property `<version.org.jboss.bom>1.0.0.Final</version.org.jboss.bom>` and replace the hard-coded version in the dependencies with `<version>${version.org.jboss.bom}</version>`
3. Add the property `<version.war.plugin>2.1.1</version.war.plugin>` and replace the hard-coded version in the dependencies and build sections.

Other changes affect the README. We've added the following metadata tags (after Author:) that provide data to display in the table on the jdf website:

Level: Beginner (Intermediate or Advanced)
Technologies: Web Services, etc.
Summary: Shows how to use Arquillian to test Web Services (or other description of the quickstart)
Target Product: EAP

Also, I'm not clear about the purpose of the openshift/WEB-INF/jboss-web.xml file. Maybe @pmuir knows?

Thanks,
Sande

Commit:
Updates to go with comments from Sande.

lnewson commented:
Hey Sande,

I've updated everything you mentioned I believe. However you mentioned not to include the remote Arquillian example and said to look at the wsat-simple example, however it contradicts what you said, so I'm assuming you meant to say removed managed and keep remote. As for the example, there really wasn't anything meaningful that could be displayed there, so I added some output to the test cases and included that in the example.

As for the Openshift folder that was a way of deploying the Application to the Root Context on Openshift but not when using the Arquillian tests. Upon thinking it out a bit more, I think that'll only cause confusion, so I've removed it and changed the OpenShift profile so that it doesn't deploy to the root context.

Cheers Lee

Commit:
Minor fix to remove a maven version warning.

sgilda commented:
Good catch! Yesterday must have been "opposite talk day" for me where I tell you the opposited of what you should do to hope you figure out the correct changes ;-). Sorry about that.

I will get your changes and try test the OpenShift deployment later today. Thanks!

sgilda commented:
Hi Lee, I realize there was one more change since you submitted this one. Again, if you look at the README for the wsat-simple quickstart, you will notice that the instructions now include deployment to EAP 6 as well as JBoss AS 7. We introduced an APPLICATION_TYPE variable to account for that. You can pretty much copy that section into your README:

Note that we use the jboss-as-quickstart@jboss.org user for these examples. You need to substitute it with your own user name.

Open a shell command prompt and change to a directory of your choice. Enter the following command, replacing APPLICATION_TYPE with `jbosseap-6.0` for quickstarts running on JBoss Enterprise Application Platform 6, or `jbossas-7` for quickstarts running on JBoss AS 7:

rhc app create -a helloworldrs -t APPLICATION_TYPE

There are a few more references to the 2 containers later on in the wsat-simple README also.

sgilda commented:
I just tested the OpenShift instructions. Under "Create the OpenShift Application", I believe this line may be a mistake:

Notice that the output also reports the URL at which the application can be accessed. Make sure it is available by typing the published url http://helloworldws-quickstart.rhcloud.com/jboss-as-helloworld-ws/ into a browser or use command line tools such as curl or wget.

Note the new "Deploy the OpenShift Application" section in the wsat-simple quickstart. The deployment doesn't happen until you push the app to OpenShift later on in the instructions. 

There is also a new section "Test the OpenShift Application". At that point you will see the "No such operation" when you access the application.

Commit:
Updates to the README

lnewson commented:
Hey Sande,

Sorry about that first URL in the openshift instructions, that was from changing the Application so it didn't deploy on the root context path and not reading the sentence. Anyways I believe I've fixed everything you mentioned, however I'm not sure what you meant by "There are a few more references to the 2 containers later on in the wsat-simple README also." in the first comment, as I couldn't find anymore references to APPLICATION_TYPE in the wsat-simple example.

Cheers Lee

sgilda commented:
Hi Lee,

I have one more minor comment. Could you add a section header "Deploy the OpenShift Application" right before the line "You can now deploy the changes to your OpenShift application using git as follows:". I know it's picky, but I'm trying to make it clear what is being done along the way at a glance and at this point you've completed the "Migrate the Quickstart Source" step.

Sorry about the cryptic reference "There are a few more references to the 2 containers later on in the wsat-simple README also." What I meant was references to the jbossas-7 and jbosseap-6.0 containers. 

For example, you have:
This command creates an OpenShift application called helloworldws and will run the application inside a `jbossas-7` container.
This needs to be changed to:
This command creates an OpenShift application called helloworldws and will run the application inside a `jbossas-7` or a `jbosseap-6.0` container.

@pmuir: Do you want to review the code one more time?

Commit:
Minor README changes.

Commit:
Commit removed old note about eap 6 not being available.

lnewson commented:
Hey Sande,

I fixed those things I believe and when re-reading it I also found an old note about EAP 6 not being available on OpenShift, so I removed that as well.

Cheers Lee

sgilda commented:
This looks great to me.

@pmuir : are you good with this now?

pmuir commented on helloworld-ws/README.md:
```diff
@@ -0,0 +1,184 @@
+What is it?
+-----------
+
+This example demonstrates the use of *JAX-WS*  in *JBoss AS 7* or *JBoss Enterprise Application Platform 6* as a simple Hello World application.
```
there is a double space after _JAX-WS_ :-)

pmuir commented on helloworld-ws/README.md:
```diff
@@ -0,0 +1,184 @@
+
+The application is deployed to <http://localhost:8080/jboss-as-helloworld-ws>. Accessing this link directly will display a SOAP "No such operation" error because it expects SOAP messages as input. You can try sending SOAP messages to the Web Service endpoint by using the Client based JUnit Tests described below.
+
+You can check that the webservice is running and deployed correctly by accessing the following URL: <http://localhost:8080/jboss-as-helloworld-ws?wsdl>. This URL will display the deployed WSDL endpoint for the Web Service.
```
I just wonder if it would be more helpful to direct people straight to the tests? Tell them they can't actually deploy the app in a meaningful way at all.

sgilda replied:
I kind of like this because you can actually see a result in the browser.
   `<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
        <soap:Body>
            <soap:Fault>
                <faultcode>soap:Server</faultcode>
                <faultstring>
                    No such operation: / (HTTP GET PATH_INFO: /jboss-as-helloworld-ws/)
                </faultstring>
            </soap:Fault>
        </soap:Body>
    </soap:Envelope>`

Lee describes it further: 
Accessing this link directly will display a SOAP “No such operation” error because it expects SOAP messages as input. You can try sending SOAP messages to the Web Service endpoint by using the Client based JUnit Tests described below.

You can check that the webservice is running and deployed correctly by accessing the following URL: http://localhost:8080/jboss-as-helloworld-ws?wsdl. This URL will display the deployed WSDL endpoint for the Web Service

I think this is useful. Especially displaying the WSDL. :-)

lnewson replied:
I did add it as I thought it would be helpful and can be somewhat useful when trying to debug (if needed). However if you guys want it removed then it doesn't bother me.

pmuir commented on helloworld-ws/pom.xml:
```diff
@@ -0,0 +1,230 @@
+		<!-- maven-compiler-plugin -->
+        <maven.compiler.target>1.6</maven.compiler.target>
+        <maven.compiler.source>1.6</maven.compiler.source>
+	</properties>
```
This isn't formatted properly. Can you reformat the file with a 4 space indent?

pmuir commented:
It would be nice to actually be able to execute the tests against the remote version deployed to openshift. We did that for the jax-rs-client quickstart, so you can probably just copy that approach. https://github.com/jboss-jdf/jboss-as-quickstart/tree/master/jax-rs-client

Commit:
Formatting Fixes as picked up by Pete Muir.

lnewson commented:
Hey Pete,

So essentially you would like to see two different type of tests? One handled by Arquillian for local testing and then another test for remote testing (since Arquillian would be useless here, due to not being able to access the management api in OpenShift).

Another possible option I saw was to use the arquillian-openshift-express adapter (http://arquillian.org/guides/testing_in_the_cloud/), however it's only in Beta and no other quickstarts are using it, so I assume that it shouldn't be used at this stage.

Cheers Lee

Commit:
Fix to deal with deploying the application properly on openshift.

sgilda commented:
@pmuir : I still vote to leave the "Access the Application" section as is.  I think it's nice to be able to view the WSDL. Do you object?

sgilda commented:
Looks good to me.

@pmuir: any other changes needed?

pmuir commented:
@lnewson my suggestion was that the setup for deploying the app to openshift should be manual, then the test executes against the manually deployed instance. Does that make sense?

@sgilda maybe it needs to be clearer that this is what you are viewing?

sgilda commented:
@pmuir: Are you thinking rather than "Access the application", maybe it should say something like "Verify the Webservice" or ""Verify the Webservice Deployment"? 

The README text explains that the errors are because SOAP expects input messages as input. Would you prefer the base URL and that paragraph be removed? 
Accessing the WSDL URL and displaying the WSDL confirms you have deployed the webservice successfully. I think that is useful. Is there something else you would want to see here?

@pmuir: I think I know what you are looking for, but none of the other quickstarts that use Arquillian tests against an OpenShift server (that I can see). Should they all be modified to do that?

pmuir commented:
Yep, having access the application return an error is odd! I would remove that bit and use the wsdl instead.

Openshift - atm the app doesn't allow the user to use it on Openshift which is odd. That's what I want to enable. The quickstart I linked earlier runs a test against Openshift.sgilda notifications@github.com wrote:@pmuir: Are you thinking rather than "Access the application", maybe it should say something like "Verify the Webservice" or ""Verify the Webservice Deployment"? 

The README text explains that the errors are because SOAP expects input messages as input. Would you prefer the base URL and that paragraph be removed? 
Accessing the WSDL URL and displaying the WSDL confirms you have deployed the webservice successfully. I think that is useful. Is there something else you would want to see here?

@pmuir: I think I know what you are looking for, but none of the other quickstarts that use Arquillian tests against an OpenShift server (that I can see). Should they all be modified to do that? 

---

Reply to this email directly or view it on GitHub:
https://github.com/jboss-jdf/jboss-as-quickstart/pull/268#issuecomment-9610032

sgilda commented:
Hi Pete, 
Removing the paragraph that describes the error accessing the URL makes total sense since you can validate the webservice using the WSDL URL (which was my biggest concern).
I also didn't realize the jax-rs-client quickstart ran the mvn test against OpenShift application. That's pretty cool.
Lee, let me know when you have finished the updates.
Thanks, Sande

lnewson commented:
Hey Pete,

I'm well aware that is what you were asking, however the guidelines state that any tests should be done in Arquillian (as per the guidelines at https://docs.jboss.org/author/display/AS72/Writing+a+quickstart). The JAX-RS client obviously breaks that rule, however it's more of a standalone application then a test so I can see how that would be allowed.

However I originally did that so that the test could be performed against any system (remote or local) and you asked that it be made into an Arquillian test. I've now done that and since as far as I've been able to find out, you can't skip the deployment phase of an Arquillian test. The deployment phase needs to be skipped if using Arquillian, because Arquillian deploys the application using the Management Interface, which isn't accessible when using Openshift (see: https://openshift.redhat.com/community/forums/openshift/jboss-as7-management-in-openshift). Therefore I have three options that I see:

1) Remove the Arquillian Test and go back to the original tests.
2) Create two different tests (the current and the old one) and configure the POM to only run a specific one for different profiles.
3) Use the arquillian-openshift-express plugin to handle the openshift setup/deployment/testing.

That is what I was asking about earlier. Either way I'm fairly busy this week and won't have time to look at it until next week anyways.

Lee

pmuir commented:
Lee, apologies, I seem to have led you round the houses here!

IMO (2) is probably the best option.

lnewson commented:
Cheers thanks Pete, I'll take a look at it over the weekend.

Commit:
Added remote client test that can be tested against openshift. Fixed
some problems pointed out in the README file.

lnewson commented:
Hey Pete and Sande,

Sorry it took me so long to get around to this again. Anyways I've added the extra test and configured the pom to run different tests based on the profile used. I've also removed the information about the error message.

One thing that I'm not really 100% sure about, was how you would have liked the remote URL specified. At the moment I've set it up as property and in the README just specified to pass the Openshift URL via the property on the command line. If you'd prefer that the instructions get the user to alter the POM then I'll change it to that.

Cheers Lee

sgilda commented:
I took a quick look at this. It builds and runs fine. I'm not sure what Pete wanted for the remote URL either, so I'll let him comment on that.

One minor issue with the README. We now put JBoss Enterprise Application Platform 6 ahead of JBoss AS 7, for example: 

This example demonstrates the use of JAX-WS JBoss Enterprise Application Platform 6 or in JBoss AS 7 as a simple Hello World application.

This is minor and I would only change it if you need to make other changes based on Pete's comments.

Commit:
Updated README.md to place the product names in the right order.

pmuir commented:
Merged
